src/reader.rs

```diff
@@ -82,17 +111,127 @@ pub fn find(needle: &ReadUntil, buffer: &str, eof: bool) -> Option<(usize, usize
                 None
             }
         }
-        &ReadUntil::Any(ref any) => {
-            for read_until in any {
-                if let Some(pos_tuple) = find(&read_until, buffer, eof) {
-                    return Some(pos_tuple);
-                }
-            }
+    }
+}
+
+pub struct Match<T> {
+    begin: usize,
+    end: usize,
+    pub interest: T,
+}
+
+impl<T> Match<T> {
+    fn new(begin: usize, end: usize, obj: T) -> Self {
+        Self {
+            begin,
+            end,
+            interest: obj,
+        }
+    }
+}
+
+pub trait Needle {
+    type Interest;
+
+    fn find(&self, buffer: &str, eof: bool) -> Option<Match<Self::Interest>>;
+}
+
+impl Needle for str {
+    type Interest = String;
+
+    fn find(&self, buffer: &str, eof: bool) -> Option<Match<Self::Interest>> {
+        buffer.find(self).and_then(|pos| {
+            Some(Match::new(
+                pos,
+                pos + self.len(),
+                buffer[..pos].to_string(),
+            ))
+        })
+    }
+}
+
+impl Needle for String {
```

Comment:
why does it need a impl both for str and for String?

Reply:
Great question, actually there's no reason in `String` implementation, as I remember I've experienced some string cast issue and just added it to not have the cast. (don't remember what it is)

And actually I was wondering at the time, might it's better not to implement the interface for existed types and create a wrapper structure e.g as I did with `NBytes(usize)` even though it could be implemented just to `usize`. I am not sure which way is better for `_string`.